File: src/main/java/ast/Child.java
package ast;

import tree.TargetTreeNode;

import java.util.ArrayList;
import java.util.List;

public class Child implements ASTNode {

	ASTNode father;
	ASTNode child;
	
	public Child(ASTNode father, ASTNode child) {
		super();
		this.father = father;
		this.child = child;
	}
	/// improve all of this shit
	@Override
	public Object execute(TargetTreeNode targetTreeNode) {
		List<TargetTreeNode> fatherList = (List<TargetTreeNode>) father.execute(targetTreeNode);
		if( fatherList.isEmpty() )
		{
			return new ArrayList<>();
		}

		List<TargetTreeNode> childList = (List<TargetTreeNode>) child.execute(targetTreeNode.moveDown());
		if( childList.isEmpty() )
		{
			return new ArrayList<>();
		}
		fatherList.addAll(childList);
		targetTreeNode.moveUp();
		return fatherList;
	}

}

File: src/main/java/ast/Child.kt
package ast

import java.util.ArrayList
import tree.TargetTreeNode

class Child(var father: ASTNode, var child: ASTNode) : ASTNode {
    // / improve all of this shit
    override fun execute(targetTreeNode: TargetTreeNode): Any? {
        val fatherList = (father.execute(targetTreeNode) as List<*>).toMutableList()
        if (fatherList.isEmpty()) {
            return ArrayList<Any>()
        }
        val childList = child.execute(targetTreeNode.moveDown()) as List<*>
        if (childList.isEmpty()) {
            return ArrayList<Any>()
        }
        fatherList.addAll(childList)
        targetTreeNode.moveUp()
        return fatherList
    }
}
